File: api/dreamcar.py
"""내 차에서 드림카까지 API."""

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field

from services import dreamcar as dc

router = APIRouter(prefix="/api/dreamcar", tags=["dreamcar"])


@router.get("/config")
def config():
    return {"questions": dc.public_questions(), "apr": dc.DEMO_APR, "terms": dc.TERMS,
            "deposit_rates": dc.DEPOSIT_RATES}


class PlateRequest(BaseModel):
    plate: str = Field(max_length=20)


@router.post("/lookup")
def lookup(req: PlateRequest):
    car = dc.lookup_car(req.plate)
    if car is None:
        raise HTTPException(400, "차량 번호 형식을 확인해주세요. 예: 123가4567")
    return car


class RecommendRequest(BaseModel):
    answers: list[int] = Field(min_length=1, max_length=10)


@router.post("/recommend")
def recommend(req: RecommendRequest):
    try:
        return dc.recommend(req.answers)
    except ValueError as e:
        raise HTTPException(400, str(e))

File: web/src/lib/dreamcar.ts
// FastAPI /api/dreamcar 클라이언트 + 할부 계산
import { apiFetch } from "@/lib/access";

export const API_URL = (process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8000").replace(/\/$/, "");
export const media = (path: string | null | undefined) => (path ? `${API_URL}${path}` : "");

export type Question = { title: string; desc: string; options: { label: string; desc: string; art: string }[] };
export type DreamConfig = { questions: Question[]; apr: number; terms: number[]; deposit_rates: number[] };
export type OwnedCar = {
  model: string;
  year: number;
  mileage: number;
  market: number;
  range_low: number;
  range_high: number;
  grade: string;
  plate: string;
};
export type ColorOption = { color: string; price: number; image: string | null };
export type TopCar = {
  model: string;
  match: number;
  tagline: string;
  reasons: string[];
  price: number;
  segment: string;
  seats: string;
  colors: ColorOption[];
};
export type Recommendation = {
  persona: { name: string; sub: string; copy: string; quote: string; art: string };
  total_models: number;
  top: TopCar[];
};

async function request<T>(path: string, body?: unknown): Promise<T> {
  const res = await apiFetch(`${API_URL}/api/dreamcar${path}`, {
    method: body ? "POST" : "GET",
    headers: { "Content-Type": "application/json" },
    body: body ? JSON.stringify(body) : undefined,
  });
  if (!res.ok) {
    let message = `서버 응답 오류 (${res.status})`;
    try {
      const j = await res.json();
      if (typeof j?.detail === "string") message = j.detail;
    } catch {}
    throw new Error(message);
  }
  return res.json() as Promise<T>;
}

export const dreamApi = {
  config: () => request<DreamConfig>("/config"),
  lookup: (plate: string) => request<OwnedCar>("/lookup", { plate }),
  recommend: (answers: number[]) => request<Recommendation>("/recommend", { answers }),
};

/** 원리금균등 월 납입금(만원)과 총 이자(만원). services/dreamcar.py monthly_installment와 같은 공식. */
export function monthlyInstallment(principalManwon: number, months: number, annualRate: number) {
  if (principalManwon <= 0) return { monthly: 0, interest: 0 };
  const principal = principalManwon * 10000;
  const r = annualRate / 100 / 12;
  const monthly = r === 0 ? principal / months : (principal * r * (1 + r) ** months) / ((1 + r) ** months - 1);
  return { monthly: monthly / 10000, interest: (monthly * months - principal) / 10000 };
}
